### ISSUE TITLE
Make Spacewalk independent of orml for Currency specification

### ISSUE THREAD

**Issue by gonzamontiel** [pendulum-chain/spacewalk#9]
Ideally, we would want that Spacewalk is independent of how users of the pallet want to implement multiple currencies in their chains.
That could be achieved by implementing our own set of traits as a layer of abstraction of those mechanisms.

As an effort in that direction, we want to minimize the dependency with orml pallets in general. So far, the spacewalk pallet is depending on `orml_tokens`, `orml_currencies` and `orml_traits`.

We want to remove the `orml_tokens` and `orml_currencies` dependencies, moving part of the related code to the `testchain`, and only depend on `orml_traits`, which is a fair compromise for the time being.

EDIT: even though `orml_traits` provides an interface for MultiCurrency that we wanna use, we don't need to depend on orml only for that spec. We need to come up with the minimum trait specification for MultiCurrency that our pallet needs and implement it. It may turn out that is really similar to https://github.com/open-web3-stack/open-runtime-module-library/blob/d42fdd6b64fbe4d234accd5d6bd862f27b97e767/traits/src/currency.rs#L18.

**Comment by gonzamontiel** [comment]
I edited the description and title of this issue to go a step further and not depend on any `orml` code for the pallet specification. That way we don't impose the users of the pallet (i.e. chains) to use a specific pallet for storing their assets.

**Comment by ashneverdawn** [comment]
We need to still keep the `orml_traits` dependency because of the `Multicurrency` trait which we need. This is because we would have to otherwise re-implement our duplicated `Multicurrency` trait for `orml_currencies::Pallet<Runtime>`, but traits can't be implemented on structs defined in an external crate, so we'd likely have to copy over a lot of code into the testchain to even attempt this.

Note that this doesn't mean that the orml pallets are required for spacewalk, it is merely the `MultiCurrency` trait that is needed.

That said, all other orml dependencies can be removed without any issue.

**Comment by gonzamontiel** [comment]
> Note that this doesn't mean that the orml pallets are required for spacewalk, it is merely the `MultiCurrency` trait that is needed. Any pallet that implements the `MultiCurrency` trait can be used in lieu of orml pallets.
 
I really don't understand this. Because following that reasoning, why we even need `orml_traits`? If the `MultiCurrency` trait was just the only thing needed, then It would have worked.
So I wouldn't say that any pallet can be used here, but only orml (or an exact copy of it, which is the same).

**Comment by ashneverdawn** [comment]
> > Note that this doesn't mean that the orml pallets are required for spacewalk, it is merely the `MultiCurrency` trait that is needed. Any pallet that implements the `MultiCurrency` trait can be used in lieu of orml pallets.
> 
> I really don't understand this. Because following that reasoning, why we even need `orml_traits`? If the `MultiCurrency` trait was just the only thing needed, then It would have worked defining it on our own pallet. So I wouldn't say that any pallet can be used here, but only orml (or an exact copy of it, which is the same).

In Rust, you cannot (yet) implement traits onto structs from external crates. `orml_currencies::Pallet<Runtime>` is a struct from an external crate. So, if you want to use your own copy of the Multicurrency trait, you can, but you'll also need to use your own copy of the struct. But this struct is also used in other pallets which also expect it to implement other traits, so you'd have to rewrite the implementations of those traits for the struct as well. If you just use the existing Multicurrency trait from the orml_traits library, then you can avoid all of that boilerplate. 

orml_traits is not a pallet, simply a lib of interfaces. Any pallet that implements the Multicurrency trait from orml_traits can be used in lieu of the orml_currencies pallet and could conceptually have very different behavior from orml_currencies, (so not an exact copy). 
Traits in rust are similar to inferfaces in other languages. Note you do need to explicitly implement a trait. Just having the matching method/signature is not enough.

**Comment by gonzamontiel** [comment]
I understand that fact. But saying that

> Any pallet that implements the Multicurrency trait from orml_traits can be used in lieu of the orml_currencies 

actually means that only `orml_currencies`/`orml_tokens` can be used there (because they are the only ones doing so), so we are actually saying that for using spacewalk you need to use those pallets or implement the whole thing yourself in your testchain. That is a restriction for the users of the pallet, which is what I was trying to circumvent. 

I'm wondering if there is any kind of Adapter that allows to plug other implementations, for example [pallet_assets](https://docs.rs/pallet-assets/latest/pallet_assets/)

**Comment by ashneverdawn** [comment]
tldr; if it's really that important to not depend on orml_traits library, we can get rid of it, but it would mean anyone using the spacewalk pallet wouldn't be able to use orml_currencies pallet as is without extra boilerplate.

**Comment by ashneverdawn** [comment]
Doesn't have to be in the actual testchain, it could be inside their custom pallet. But if they don't need to implement orml_traits Multicurrency, they'd still have to implement spacewalk's MultiCurrency. So if they aren't using orml, there's no difference to them either way. I feel like everyone using a common set of traits helps with compatibility and reducing wrapper boilerplate around stuff generally. But now that you mention it, it's true that there aren't that many pallets currently using the orml MultiCurrency trait.

**Comment by gonzamontiel** [comment]
Honestly, I think this is a common problem among parachains: how to implement an interoperable way of storing custom assets and transfer them. And even though we're not outside of that problem, it is fair to say that we _need_ to pick one of the available mechanisms out there for storing multiple assets. The one we chose will have certain compatibility with existing chains. So for simplicity I would consider this closed _as is_, meaning only depending on `orml_traits` for `MultiCurrency`. 

Also in favor of that: apparently interlay it's using the same trait in [their implementation of currency](https://github.com/interlay/interbtc/blob/f9abdebf5c1dca9288a39c706b8414ab8dac4100/crates/currency/src/amount.rs#L10).

I will then review the PR!

**Comment by TorstenStueber** [comment]
I would find it preferable if Spacewalk would only use the standard `Currency` trait (as defined in "frame-support").

In the testchain and in pendulum we currently use orml `MultiCurrency`. Shouldn't it be possible to write an adapter from `MultiCurrency` to frame's `Currency` (which is actually called `PalletCurrency` in orml)?

Something like a function (MultiCurrency, currency_id) -> impl Currency?

Seems like that would not work because the function would need to return a value that keeps track of the `currency_id`, e.g., a struct with the `currency_id`. This can't work because the `Currency` trait seems to me to be surprisingly non-idiomatic: for example it has a function `fn total_issuance() -> Self::Balance` which returns the current totally issued amount. Usually such a function would need to have the signature `fn total_issuance(&self) -> Self::Balance` in order to return some dynamic internal value. It's quite magical that `&self` is not required here and looks like it would need to access some global static value. I think what it actually does is to call an external function that accesses the storage outside the WASM runtime and for that reason there is no danger of data races.

This is unfortunate as it makes it impossible to return an `impl Currency` in the function above as for example the function `total_issuance` that needs to be implemented on that value would need to access the `currency_id`, but it can't due to the missing `&self`.

**Comment by TorstenStueber** [comment]
My previous comment argues that there is a clear distinction with "Mono Currencies" and "Multi Currencies" and one cannot be mapped to the other.

I realized that Substrate defines a standard for _Multi Currencies_, it is [`fungibles`](https://github.com/paritytech/substrate/blob/master/frame/support/src/traits/tokens/fungibles.rs) (not [`fungible`](https://github.com/paritytech/substrate/blob/master/frame/support/src/traits/tokens/fungible.rs) – that would be a mono currency).

So I suggest that we use `fungibles::Inspect + fungibles::Mutate + fungibles::Transfer + fungibles::Create` as the required type. `Inspect` as basic currency functionality, `Mutate` for burning and minting, `Transfer` for token transfer and `Create` to create new assets.

**Comment by ashneverdawn** [comment]
> My previous comment argues that there is a clear distinction with "Mono Currencies" and "Multi Currencies" and one cannot be mapped to the other.
> 
> I realized that Substrate defines a standard for _Multi Currencies_, it is [`fungibles`](https://github.com/paritytech/substrate/blob/master/frame/support/src/traits/tokens/fungibles.rs) (not [`fungible`](https://github.com/paritytech/substrate/blob/master/frame/support/src/traits/tokens/fungible.rs) – that would be a mono currency).
> 
> So I suggest that we use `fungibles::Inspect + fungibles::Mutate + fungibles::Transfer + fungibles::Create` as the required type. `Inspect` as basic currency functionality, `Mutate` for burning and minting, `Transfer` for token transfer and `Create` to create new assets.

Just updating here from what you guys discussed on Slack. Since it turns out that orml's MultiCurrency has actually become more 'standard' or common than using the built in fungibles, then MultiCurrency became a viable choice again. 

Either of these would fit our needs. From the discussion, seems like orml's MultiCurrency is what we will end up using. Hopefully it gets merged into substrate in the future.
